Add subctl upgrade command

Jaanki opened the pull request:
See individual commits for description

Epic: https://github.com/submariner-io/enhancements/issues/183

Jaanki committed:
Add GHA to test subctl upgrades

This commit adds a GHA that does following:

1. Deploys the environment using latest images
2. Builds the new subctl
3. Run subctl upgrade command to upgrade the environment to the PR's
   version
4. Run e2e tests

Epic: https://github.com/submariner-io/enhancements/issues/183

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

Jaanki committed:
Add upgrade logic

1. Upgrade Broker to specified version by re-deploying it
2. Upgrade Submariner/Service discovery to specified version by redeploying
it with the specified version

Move getBroker function out of recover_broker_info.go file because that
is excluded while building subctl and hence upgrade.go can't find that
function.

Epic: https://github.com/submariner-io/enhancements/issues/183

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

Jaanki committed:
Upgrade subctl

Upgrade subctl to the user-specified version by downloading
and installing the specified subctl version using the recommended
curl method.

Epic: https://github.com/submariner-io/enhancements/issues/183

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

Jaanki commented:
Upgrade command jb fails because the upgrade job from shipyard that it depends on, is buggy.

skitt committed:
Upgrades: handle devel, rework version checks

If the requested versions already match, skip upgrading subctl.
If a specific version isn't requested, download the latest release.
Only show status information if a subctl upgrade is actually
necessary, since everything else is instantaneous.
Don't suggest that requesting an older version of subctl will result
in an early exit; the installed version of Submariner may be older
still, and therefore an upgrade might be needed anyway.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt committed:
Upgrades: restart subctl if it was updated

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt committed:
Upgrades: retrieve the latest tag from GitHub

This avoids downloading subctl when it's already the latest version.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt committed:
Upgrades: allow the broker to be on a shared cluster

During upgrades, finding a broker shouldn't exclude the rest of the
upgrade process on the cluster being processed.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+}
+
+func init() {
+	upgradeCmd.Flags().StringVar(&to, "to", "", "the version of Submariner to upgrade to")
```
```suggestion
	upgradeCmd.Flags().StringVar(&upgradeVersion, "to-version", "", "the version of Submariner to which to upgrade")
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+
+// upgradeSubctl upgrades the local copy of subctl, if necessary.
+// Returns the path to the upgraded subctl if subctl was upgraded, nil if it wasn't.
+func upgradeSubctl(status reporter.Interface) (*string, error) {
```
Minor thing but I think it's more idiomatic to return an empty string rather than nil, ie an empty string is equivalent to nil.

```suggestion
func upgradeSubctl(status reporter.Interface) (string, error) {
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+
+	brokerObj, brokerFound, err := getBroker(clusterInfo.RestConfig, constants.DefaultBrokerNamespace)
+	if err != nil {
+		return err
```
So it ends the status.

```suggestion
		return status.Error(err, "")
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+		// Role updates are part of Broker redeploy
+		err = upgradeBroker(ctx, clusterInfo, status, brokerObj.Namespace, brokerObj.Spec)
+		if err != nil {
+			return status.Error(err, "Error upgrading Broker")
```
The status should be updated by `upgradeBroker` since it starts it.

```suggestion
			return err
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+
+	// Upgrade Operator if deployed
+	if err := upgradeOperator(ctx, clusterInfo, repository, debug, imageOverride, status); err != nil {
+		return status.Error(err, "Error upgrading Operator")
```
The status should be updated by `upgradeOperator` since it starts it.

```suggestion
		return err
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+
+	// Upgrade Submariner
+	if clusterInfo.Submariner != nil {
+		status.Start("Found Submariner components. Upgrading it to %s", to)
```
```suggestion
		status.Start("The Connectivity component is installed. Upgrading to version %s", to)
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+
+		err := deploy.SubmarinerFromSpec(ctx, clusterInfo.ClientProducer.ForGeneral(), &clusterInfo.Submariner.Spec)
+		if err != nil {
+			return status.Error(err, "Error upgrading Submariner")
```
```suggestion
			return status.Error(err, "Error upgrading the Connectivity component")
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+			return status.Error(err, "Error upgrading Submariner")
+		}
+
+		status.Success("Submariner successfully upgraded")
```
We don't need an explicit success message as ending the status indicates success.

```suggestion
		status.End()
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+
+	// Upgrade Service discovery
+	if clusterInfo.ServiceDiscovery != nil {
+		status.Start("Found Service Discovery components. Upgrading it to %s", to)
```
```suggestion
		status.Start("The Service Discovery component is installed. Upgrading to version %s", to)
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+			return status.Error(err, "Error upgrading Service Discovery")
+		}
+
+		status.Success("Service discovery successfully upgraded.")
```
```suggestion
		status.End()
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+func upgradeBroker(ctx context.Context, clusterInfo *cluster.Info, status reporter.Interface, namespace string,
+	brokerSpec v1alpha1.BrokerSpec,
+) error {
+	status.Start("Found Broker installed. Upgrading it to %s", to)
```
```suggestion
	status.Start("The Broker is installed. Upgrading to version %s", to)
	status.End()
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+	}
+
+	if err := deploy.Deploy(ctx, options, status, clusterInfo.ClientProducer); err != nil {
+		return err //nolint:wrapcheck // No need to wrap here
```
```suggestion
		return status.Error(err, "Error upgrading the Broker")
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+		return err //nolint:wrapcheck // No need to wrap here
+	}
+
+	status.Success("Broker successfully upgraded.")
```
Not needed.

```suggestion
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+func upgradeOperator(ctx context.Context, clusterInfo *cluster.Info, repository string, debug bool, imageOverride map[string]string,
+	status reporter.Interface,
+) error {
+	status.Start("Checking if Operator is deployed")
```
```suggestion
	status.Start("Checking if the Operator is installed")
```

Note that we use the word "installed" elsewhere so we should be consistent (or use the word "deployed" everywhere).

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+func upgradeSubmariner(clusterInfo *cluster.Info, _ string, status reporter.Interface) error {
+	ctx := context.TODO()
+
+	status.Start("Starting upgrade process")
+	defer status.End()
```
```suggestion
	status.Start("Checking if the Broker is installed")
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+	brokerObj, brokerFound, err := getBroker(clusterInfo.RestConfig, constants.DefaultBrokerNamespace)
+	if err != nil {
+		return err
+	}
```
```suggestion
	}
	
	status.End()
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+		return status.Error(err, "Error retrieving Operator deployment")
+	}
+
+	if operatorFound {
+		status.Success("Operator deployed. Upgrading it")
+
+		err = deploy.Operator(
+			ctx, status, repository, to, imageOverride,
+			clusterInfo.ClientProducer, debug)
+		if err != nil {
+			return err //nolint:wrapcheck // No need to wrap here
+		}
+
+		status.Success("Operator successfully upgraded")
+	}
```
I think it's extraneous to have a function, `deploy.Operator`,  that takes 7 params that essentially wraps one line of code, especially if it's only used in one place. I think we should just inline it.

```suggestion
	status.Start("The Operator is installed. Upgrading to version %s", to)
        status.End()
        
        repositoryInfo := image.NewRepositoryInfo(repository, to, imageOverride)

	err := operator.Ensure(
		ctx, status, clientProducer, constants.OperatorNamespace, repositoryInfo.GetOperatorImage(), debug)
	if err != nil {
	    return status.Error(err, "Error upgrading the Operator")
	}
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+	status.Start("Checking if Operator is deployed")
+	defer status.End()
+
+	operatorFound, err := deploy.IsOperator(ctx, clusterInfo.ClientProducer.ForKubernetes())
+	if err != nil {
+		return status.Error(err, "Error retrieving Operator deployment")
+	}
```
The name `deploy.IsOperator` is a bit awkward. Perhaps `deploy.HasOperator` would be better but it basically just wraps an API server call so we may as well inline it here

```suggestion
        _, err := clusterInfo.ClientProducer.ForKubernetes().AppsV1().Deployments(constants.OperatorNamespace).Get(ctx, names.OperatorComponent, metav1.GetOptions{})
        if resource.IsNotFoundErr(err) {
            return nil
        }
        
	if err != nil {
		return status.Error(err, "Error retrieving the Operator deployment")
	}
```

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,271 @@
+		return nil, status.Error(err, "Error upgrading subctl")
+	}
+
+	status.Success("Upgraded and installed subctl version: %s", to)
```
Not needed.

```suggestion
```

skitt committed:
Upgrades: make version-handling consistent

For testing, we need to be able to specify the operator and Submariner
versions separately; allow that using hidden parameters.

Encapsulate all the various upgrades, and ensure that status
information is handled consistently. In particular, success doesn't
need a separate message, status.End() marks the operation as
successful.

Upgrading the broker upgrades the operator, so skip the explicit
operator upgrade if the broker has been upgraded.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

tpantelis commented on cmd/subctl/upgrade.go:
```diff
@@ -0,0 +1,308 @@
+
+	operatorDeployment, err := clusterInfo.ClientProducer.ForKubernetes().AppsV1().Deployments(constants.OperatorNamespace).
+		Get(ctx, names.OperatorComponent, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return status.Error(err, "Error retrieving Operator deployment")
+	}
+
+	if operatorDeployment == nil {
+		return nil
+	}
```
Is it guaranteed that `operatorDeployment` will be nil if not found? I think it's clearer to key off of `IsNotFound`.

```suggestion
	if errors.IsNotFound(err) {
		return nil
	}

	if err != nil  {
		return status.Error(err, "Error retrieving Operator deployment")
	}
```

skitt replied:
Right, given the implementation, probably not.

tpantelis reviewed (approved):
Looks good. Minor comment if you re-spin.

sridhargaddam commented on .github/workflows/upgrade-subctl.yml:
```diff
@@ -0,0 +1,45 @@
+
+on:
+  pull_request:
+    types: [ready_for_review, opened, reopened, synchronize, converted_to_draft, labeled]
```
Any specific reason why we want to run this job when a PR is converted to draft?

skitt replied:
No, I suspect this was copied from `upgrade-e2e.yml` where it is paired with an `if` clause in the job.

sridhargaddam commented on cmd/subctl/broker.go:
```diff
@@ -0,0 +1,55 @@
+func getBroker(ctx context.Context, config *rest.Config, namespace string) (*v1alpha1.Broker, bool, error) {
+	brokerClientProducer, err := client.NewProducerFromRestConfig(config)
+	if err != nil {
+		return nil, false, errors.Wrap(err, "Error creating broker client Producer")
```
nit:

```suggestion
		return nil, false, errors.Wrap(err, "error creating broker client Producer")
```

skitt committed:
Merge branch 'devel' into upgrade
